### ISSUE TITLE
Timeout exploration technique

### ISSUE THREAD

**Issue by zardus** [angr/angr#682]
A common request is an easy (and clean) way to timeout symbolic exploration. I good beginner project for someone to get involved in angr would be to create this as an exploration technique.

**Comment by VitorFalcao** [comment]
@zardus Hi, I would like to start contributing to the project. This issue seems very interesting. Could you please elaborate more on the issue?

**Comment by rhelmot** [comment]
Hi, thanks for your interest!

You probably know that most emulation and analysis with angr happens using a simulation manager. Lots of common scenarios involve setting up some search strategy and telling the simulation to run until some termination condition, such as finding a state at a given address, or finding a crash. One termination condition that people want to be able to use but isn't readily available is the ability to add a timeout, or that exploration should stop after a given number of seconds.

Simulation managers can have their behavior extended by writing an [Exploration Technique](http://angr.io/api-doc/angr.html#angr.exploration_techniques.ExplorationTechnique), which is a subclass of the ExplorationTechnique base class with some methods overridden. Some examples of this can be found [here](https://github.com/angr/angr/tree/master/angr/exploration_techniques). The goal of this issue is to have there be an exploration technique which causes emulation to halt after a timeout.

Probably the difficult part of this will be determining when to _start_ the timer, since there is no callback that a technique class can use to be informed of when a long-running execution is started. If you think that that this would be a valuable thing to have, you could add that to the ExplorationTechnique base class and integrate it with the simulation manger, or you could come up with a different solution using the tools already available.

Good luck!

**Comment by VitorFalcao** [comment]
@rhelmot Thank you very much for the explanation. I am currently learning how the code works and then I will try to come up with a solution.

**Comment by humeafo** [comment]
Hi @VitorFalcao the following is one implementation I wrote  some time ago, maybe you can take it as a reference, sorry for my code was not throughly tested, so just test it.

The tech provided two kinds of timeout, one is per-state timeout and another is total angr instance timeout.
the start time maybe not so accurate so a better way maybe to do as @rhelmot suggests, about the state timeout, current implementation perhaps is not good, you perhaps should consider using simgr's step callback to do per-state time-counting and timeout checks if this functionality is needed at all...

I'm new to angr too, so please keep doubting every line of my codes...

`
import angr
from angr import ExplorationTechnique
import sched, time

import logging
l = logging.getLogger('angr.rlimiters')

PRUNED_SLOTNAME = "pruned"

class TimeLimiter(ExplorationTechnique):
  """
  limit simulation manager's total wall time or state's runtime in seconds
  """
  def __init__(self, max_time=24*3600, max_state_time=None):
    super(TimeLimiter, self).__init__()
    self.max_time = max_time
    self.max_state_time = max_state_time
    self.scheduler = sched.scheduler()

    self.all_timeout = False
    self.timeout_states = set()

  def _on_time_consumed(self):
    self.all_timeout = True
  
  def _on_engine_process_pre(self, state):
    state.globals['__tl_pre_time'] = time.time()

  def _on_engine__process_post(self, state):
    ctag = time.time()
    ss = state.inspect.sim_successors

    for s in ss.flat_successors:
      stag = s.globals.get('__tl_pre_time', ctag)
      cstime = ctag - stag
      if cstime > 0:
        kt = '__tl_stime'
        if kt not in s.globals:
          s.globals[kt] = 0
        s.globals[kt] += cstime
        if s.globals[kt] > self.max_state_time:
          self.timeout_states.add(s)

  def setup(self, simgr):
    if self.max_time:
      self.scheduler.enter(self.max_time, 1, self._on_time_consumed)
      self.scheduler.run()

    if self.max_state_time:
      state.inspect.b('engine_process', when=angr.BP_BEFORE, action=self._on_engine_process_pre)
      state.inspect.b('engine_process', when=angr.BP_AFTER, action=self._on_engine__process_post)

  def filter(self, state):
    if self.all_timeout:
      return PRUNED_SLOTNAME

    if state in self.timeout_states:
      self.timeout_states.remove(state)
      return PRUNED_SLOTNAME
    return None

`

**Comment by humeafo** [comment]
I just noted angr already had a state block length limiter, so the per-state timeout is indeed redundant.

**Comment by humeafo** [comment]
I updated another version after lengthlimiter, there maybe some more to improve:

    def step(self, pg, stash, **kwargs):
        pg = pg.step(stash=stash, **kwargs)
        pg.move('active', '_DROP' if self._drop else 'cut', self._filter)
        return pg

maybe it's better to rename pg to simgr, and replace 'active' with stash, however, this is just typos that not affect results.

**Comment by danse-macabre** [comment]
@humeafo Let me disagree with you on this matter. Evaluation of a severely complex state could take a long time, and you certainly don't want to suddenly know, that the evaluation has timed out _after_ the precious time has been spent.

One possible way to fix this problem is to use a simple `timeout` decorator:

```
import signal


class TimeoutError(Exception):
    pass


def is_debugging():    
    pass


class timeout:

    def __init__(self, seconds=1, error_message='Timeout'):
        self.seconds = seconds
        self.error_message = error_message

    def handle_timeout(self, signum, frame):
        if not is_debugging():  # Timeouts should be disabled while debugging.
            raise TimeoutError(self.error_message)
        print 'TimeoutError was not raised due to debug is ongoing!'

    def __enter__(self):
        signal.signal(signal.SIGALRM, self.handle_timeout)
        signal.alarm(self.seconds)

    def __exit__(self, type, value, traceback):
        signal.alarm(0)
```

Note, that this example `timeout` implementation is Linux-only. For cross-platform solution consider using the [timeout-decorator](https://pypi.python.org/pypi/timeout-decorator).

In context of exploration technique it could be used like this:

```
class TimeLimitedExecution(ExplorationTechnique):

    def __init__(self, time_limit):
        super(TimeLimitedExecution, self).__init__()
        self.time_limit = time_limit

    def step_state(self, state, **kwargs):
        try:
            with timeout(self.time_limit):
                return super(TimeLimitedExecution, self).step_state(state, **kwargs)
        except TimeoutError:
            return {'timeout': [state]}  # Move state to 'timeout' stash
```

**Comment by humeafo** [comment]
It’s doable, but counting time slice of every path on every step consume time too, block length limit is a bit light, and I think it’s adequate in most cases I encountered to prevent bad path.

The super line in step-state may not work, I am not sure now, worth a source check.step of tech may not be the right place too,and in engine-process inspect seems a bit ugly...

**Comment by humeafo** [comment]
Following is my refined-tested version, with per-state timeout support, you decide to have it or not:

`import angr, time, threading

import logging
l = logging.getLogger('angr.exploration_techniques.timelimiter')

BEGIN_TIMEKEY = '__tl_begin_time'
STATE_TIMEKEY = '__tl_state_time'

class TimeLimiter(angr.ExplorationTechnique):
    '''
    Time limiter on angr instance or per state in seconds.
    '''

    def __init__(self, max_time, max_state_time=None, drop=True):
        super(TimeLimiter, self).__init__()
        self._max_time = max_time
        self._max_state_time = max_state_time

        self._slot_name = '_DROP' if drop else 'cut'
        self._all_timeouted = False
        self._executor = threading.Thread(target=self._time_monitor)

        if self._max_time is None or self._max_time <= 0:
            l.debug('timelimiter instance timeout is disabled: max time {} is illegal, should be in seconds'.format(self._max_time))
            self._max_time = None
        if self._max_state_time is None or self._max_state_time <= 0:
            l.debug('timelimiter state time limiter is disabled: max time {} is illegal, should be in seconds'.format(self._max_state_time))
            self._max_state_time = None

    def _on_engine_process_pre(self, state):
        state.globals[BEGIN_TIMEKEY] = time.time()

    def _on_engine_process_post(self, state):
        ss = state.inspect.sim_successors
        if len(ss.flat_successors) > 0:
            ctag = time.time()

            for s in ss.flat_successors:
                stag = s.globals.get(BEGIN_TIMEKEY, ctag)
                process_time = ctag - stag
                if process_time > 0:
                    if STATE_TIMEKEY not in s.globals:
                        s.globals[STATE_TIMEKEY] = process_time
                    else:
                        s.globals[STATE_TIMEKEY] += process_time

    def _time_monitor(self):
        time.sleep(self._max_time)
        self._all_timeouted = True
        l.warning('max time limit {} seconds reached'.format(self._max_time))

    def setup(self, simgr):
        if self._max_time:
            self._executor.daemon=True
            self._executor.start()

        if self._max_state_time:
            for state in simgr.active:
                state.inspect.b('engine_process', when=angr.BP_BEFORE, action=self._on_engine_process_pre)
                state.inspect.b('engine_process', when=angr.BP_AFTER, action=self._on_engine_process_post)

    def filter(self, state):
        if self._all_timeouted:
            return self._slot_name

        if self._max_state_time is not None and state.globals.get(STATE_TIMEKEY, 0) >= self._max_state_time:
            l.warning('killing state {} for max time limit {} seconds reached'.format(state, self._max_state_time))
            return self._slot_name
        return None
`

**Comment by danse-macabre** [comment]
@humeafo I think, we should definitely consult @zardus or @rhelmot on this matter.

**Comment by humeafo** [comment]
@danse-macabre Yes, I can't agree more...

**Comment by rhelmot** [comment]
The breakpoint stuff is definitely not a good idea, since there's no good way to make sure that you get to tear down setup stuff that you do to each of the states you're working with.

I'm definitely not following the conversation on per-state timeouts - the only thing I think this issue is trying to address is being able to set a global timeout for the simulation manager. You can accomplish this with a signal by tracking yourself how much time has elapsed since the timer has started, and then setting a signal-based timer to wake up after `timeout - time_elapsed` seconds. I think a per-state timeout with a `timed_out` stash would be better suited for a different technique entirely.

**Comment by zardus** [comment]
I agree, the alarm-based method of overriding `step_state` is better than the breakpoint method. I think that an exploration technique that supports this and a general timeout would be perfect. Something like:

```
simgr.use_technique(angr.exploration_techniques.Timeout(per_state=10, overall=100))
```

When given both, it could use the alarm per-state, and track a counter overall. When only given overall, it can use the alarm overall.

@rhelmot, I think this is a very good use-case for nestable `step_path` callbacks for oties. No reason this shouldn't be usable with veritesting and so forth. However, for that, I think we'll need to pass the simgr into `step_path` as well.

Also, I keep saying `alarm` above, but we should really use `TimeoutDecorator`, since that claims to work with threading (though I don't know how).

**Comment by humeafo** [comment]
Is the TimeoutDecorator specified is this? https://github.com/pnpnpn/timeout-decorator/blob/master/timeout_decorator/timeout_decorator.py, it used multiprocessing and queue to exec target function if signal alarm is not used, some overhead is involved, especially on win32 with no forking support. signal alarm is not portable and has thread issue.
In fact, even with simple time.time() arith, it would only take several seconds(on my machine 4s for cpython) for 10000000 calls, I think it's negligible for 10000000 steps...

**Comment by zardus** [comment]
Yes, but with a manual timing-based method, there's no protection against Unicorn or Z3 hanging during operation. But that's probably ok for a first run at this.

**Comment by humeafo** [comment]
If z3 or Unicorn blocking is considered, maybe a separate thread(thread.timer) would be a viable solution for the overall time limit, the only shortcoming of this scheme is that whole process must be terminated when timeout because of hanging.

**Comment by zardus** [comment]
If we're going to go that route, I would highly recommend to just use the `TimeoutDecorator` and let it handle the implementation details.

**Comment by ambiso** [comment]
Is development of this still active?

**Comment by rhelmot** [comment]
Yes - none of the discussion above ever yielded any code IIRC. One important thing is that `step_state` has become nestable since then, so this should be much easier now!

**Comment by edmcman** [comment]
I've been working on this feature for myself and it's not as easy as letting `TimeoutDecorator` handle all the implementation details.  There are two implementations.

The first implementation uses signals.  Unfortunately, it doesn't seem to be able to interrupt a FFI function call, e.g., z3.  Since most of the time z3 is the culprit, this isn't very useful.

The other implementation spawns a new thread to execute the timed code.  Something funky is going wrong here too.  I am still investigating, but an initial guess is that hooklib is not playing nicely with pickling.

**Comment by edmcman** [comment]
My guess about the hooks was completely wrong.  It turns out that SimSegFaultException can't be pickled, which was causing a very obscure error. #1986

**Comment by edmcman** [comment]
Even after fixing the pickling issues, it turns out that pickling the simgr is pretty slow.  I think the best option might be to use `use_signals=True`, set the solver timeout, and then hope that covers most of the situations.

**Comment by ltfish** [comment]
> it turns out that pickling the simgr is pretty slow

This is the same problem as I have seen in angr.distributed (yet to be merged). I think some careful profiling will alleviate the problem, but I feel a better way going forward is to avoid pickling and switch to some other serialization techniques.

**Comment by aa-software2112** [comment]
Hello! 

Has any progress been made related to this issue? This would be a huge contribution in terms of symbolic execution. Some other solution I've found online require training a NN model to predict how long the solver would take on the upcoming constraints, but the overhead is painful.

I'm trying to build a speedy symbolic execution system atop angr, and the only thing (so far) slowing it down is the inability for me to control execution time _exactly_ (i.e. I can set a global timer to 1 minute, but if the solver hangs for 10 minutes on a step(), the global timer is only checked _after_ the constraints are solved).

**Comment by ltfish** [comment]
I hope this is not a stupid question to ask, but do z3 APIs have a timeout setting? If not, we probably want to introduce timeout into Z3 (instead of hacking things in the python land).

**Comment by aa-software2112** [comment]
> I hope this is not a stupid question to ask, but do z3 APIs have a timeout setting? If not, we probably want to introduce timeout into Z3 (instead of hacking things in the python land).

If you do a google search on z3 timeouts, it seems z3 takes a timeout parameter at some point

**Comment by edmcman** [comment]
I'll post some code I have for this that works pretty well. It uses the timeout decorator above and sets the z3 timeout through the clarity backend. (I'm on my phone right now)

**Comment by aa-software2112** [comment]
Checkout this--> http://angr.io/api-doc/claripy.html#claripy.backends.Backend
Notice there is a solver(timeout=None) function at the base class of each Claripy backend

**Comment by aa-software2112** [comment]
> I'll post some code I have for this that works pretty well. It uses the timeout decorator above and sets the z3 timeout through the clarity backend. (I'm on my phone right now)

You posted that just as I found that very Claripy capability, looking forward to seeing your solution!

**Comment by edmcman** [comment]
Here's what I'm using.  

[measure_time.zip](https://github.com/angr/angr/files/6672334/measure_time.zip)

It's been a while, but I don't think the timeout argument that you referenced worked for me.  I had to reach into the Z3 internals via `state.solver._solver.timeout = int(self.step_timeout * 1000)`.  But I could be misremembering things, or it may have changed since then.

**Comment by aa-software2112** [comment]
> Here's what I'm using.
> 
> [measure_time.zip](https://github.com/angr/angr/files/6672334/measure_time.zip)
> 
> It's been a while, but I don't think the timeout argument that you referenced worked for me. I had to reach into the Z3 internals via `state.solver._solver.timeout = int(self.step_timeout * 1000)`. But I could be misremembering things, or it may have changed since then.

Thank you, I will be testing this out next week

**Comment by aa-software2112** [comment]
@edmcman 
For the context of my work, `state.solver._solver.timeout` did the trick for me - I didn't need an extra signal work. However I do have an additional question I'm hoping you can help me with.

In the case where there _is_ a solver timeout triggered, how do we actually know if the timeout occurred? I don't seem to get an exception, or any return value indicating this. Others (https://github.com/angr/claripy/issues/41) have suggested checking the check_satisfiability(...) or satisfiable(...). Given the API for these two (below), it really doesn't seem like they would tell us if the solved result returned in the face of a timeout is valid or not... 

![image](https://user-images.githubusercontent.com/22150653/122596429-d7f70180-d037-11eb-85a3-04c4b1846a93.png)

Oddly enough, even when the timeout is triggered, a solution is returned to me (I know this because the block I'm symbolically executing usually takes 30 seconds+, and with a 5s timeout it ends in 5 seconds).

**Comment by edmcman** [comment]
For me the timeout wrapper throws an exception as soon as z3 times out and returns.

**Comment by twizmwazin** [comment]
Fixed by #3153
